tangelo/toolboxes/qubit_mappings/combinatorial.py

```diff
@@ -0,0 +1,234 @@
+        for sigma_beta, int_beta in basis_set_beta.items():
+            # Alternate ordering (like FermionOperator in openfermion).
+            sigma = tuple(sorted([2*sa for sa in sigma_alpha] + [2*sb+1 for sb in sigma_beta]))
+            unique_int = (int_alpha * n_choose_alpha) + int_beta
+            basis_set[sigma] = unique_int
+
+    # Compact Hamiltonian initialization to 2^n * 2^n matrices.
+    h_c = np.zeros((2**n, 2**n), dtype=complex)
+
+    # Check what is the effect of every term.
+    for term, coeff in ferm_op_chemist.terms.items():
+        # Core term.
+        if not term:
+            continue
+
+        # Get the effect of each operator to the basis set items.
+        for b, unique_int in basis_set.items():
+
+            new_state, phase = one_body_op_on_state(term[-2:], b)
+
+            if len(term) == 4 and new_state:
+                new_state, phase_two = one_body_op_on_state(term[:2], new_state)
+                phase *= phase_two
+
+            if not new_state:
+                continue
+
+            new_unique_int = basis_set[new_state]
+            h_c[unique_int][new_unique_int] += phase * coeff
+
+    return h_to_qubitop(h_c, n) + ferm_op_chemist.constant
+
+
+def basis(M, N):
+    """Function to construct the combinatorial basis set, i.e. a basis set
+    respecting the number of electrons and the total spin.
+
+    Args:
+        M (int): Number of spatial orbitals.
+        N (int): Number of alpha or beta electrons.
+
+    Returns:
+        OrderedDict: Lexicographically sorted basis set, mapping electronic
+            configuration to unique integers.
+    """
+
+    mapping = [(sigma, f(sigma, M)) for sigma in itertools.combinations(range(M), N)]
+    return OrderedDict(mapping)
+
+
+def f(sigma, M):
+    """Function to map an electronic configuration to a unique integer, as done
+    in arXiv.2205.11742 eq. (14).
+
+    Args:
+        sigma (tuple of int): Orbital indices where the electron are in the
+            electronic configuration.
+        M (int): Number of modes, i.e. number of spatial orbitals.
+
+    Returns:
+        int: Unique integer for the input electronic state.
+    """
+
+    # Equivalent to the number of electrons.
+    N = len(sigma)
+
+    # Eq. (14) in the reference.
+    terms_k = [comb(M - sigma[N - 1 - k] - 1, k + 1) for k in range(N)]
+    unique_int = comb(M, N) - 1 - np.sum(terms_k)
+
+    return int(unique_int)
+
+
+def one_body_op_on_state(op, state_in):
+    """Function to apply a a^{\dagger}_i a_j operator as described in Phys. Rev.
+    A 81, 022124 (2010) eq. (8).
+
+    Args:
+        op (tuple): Operator, written as ((qubit_i, 1), (qubit_j, 0)), where 0/1
+            means anhilation/creation on the specified qubit.
+        state_in (tuple): Electronic configuration described as tuple of
+            spinorbital indices where there is an electron.
+
+    Returns:
+        tuple or 0: Resulting state with the same form as in the input state.
```

Comment:
it does not look like this function can just return 0. It seems to always return a tuple.